homeassistant/components/switcher_kis/config_flow.py

```diff
@@ -2,9 +2,119 @@
 
 from __future__ import annotations
 
-from homeassistant.helpers import config_entry_flow
+from collections.abc import Mapping
+import logging
+from typing import Any
 
-from .const import DOMAIN
-from .utils import async_has_devices
+from aioswitcher.bridge import SwitcherBase
+from aioswitcher.device.tools import validate_token
+import voluptuous as vol
 
-config_entry_flow.register_discovery_flow(DOMAIN, "Switcher", async_has_devices)
+from homeassistant.config_entries import ConfigEntry, ConfigFlow, ConfigFlowResult
+
+from .const import CONF_TOKEN, CONF_USERNAME, DOMAIN
+from .utils import async_discover_devices
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SwitcherFlowHandler(ConfigFlow, domain=DOMAIN):
+    """Handle Switcher config flow."""
+
+    VERSION = 1
+
+    entry: ConfigEntry | None = None
+    username: str | None = None
+    token: str | None = None
+    discovered_devices: dict[str, SwitcherBase] = {}
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the start of the config flow."""
+        self.discovered_devices = await async_discover_devices()
+
+        return self.async_show_form(step_id="confirm")
+
+    async def async_step_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle user-confirmation of the config flow."""
+        if len(self.discovered_devices) == 0:
+            return self.async_abort(reason="no_devices_found")
+
+        for device_id, device in self.discovered_devices.items():
+            if device.token_needed:
+                _LOGGER.info("Device with ID %s requires a token", device_id)
+                return await self.async_step_credentials()
+        return await self._create_entry()
+
+    async def async_step_credentials(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the credentials step."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            self.username = user_input.get(CONF_USERNAME)
+            self.token = user_input.get(CONF_TOKEN)
+
+            token_is_valid = await validate_token(
+                user_input[CONF_USERNAME], user_input[CONF_TOKEN]
+            )
+            if token_is_valid:
+                return await self._create_entry()
+            errors["base"] = "invalid_auth"
+        else:
+            user_input = {}
+
+        schema = {
+            vol.Required(CONF_USERNAME, default=user_input.get(CONF_USERNAME, "")): str,
+            vol.Required(CONF_TOKEN, default=user_input.get(CONF_TOKEN, "")): str,
+        }
```

Comment:
Please move the schema as a constant outside of the class (since you use it in more than one place), example:
https://github.com/home-assistant/core/blob/50a46933f62c2495cdc5e781ee13aa872c98036b/homeassistant/components/shelly/config_flow.py#L60-L65

Reply:
It is done.